lib/streamlit/errors.py

```diff
@@ -133,3 +136,249 @@ def __init__(self, module_name, *args):
             "installed."
         )
         super().__init__(message, *args)
+
+
+class LocalizableStreamlitException(StreamlitAPIException):
+    def __init__(self, message: str, **kwargs):
+        super().__init__((message).format(**kwargs))
+        self._exec_kwargs = kwargs
+
+    @property
+    def exec_kwargs(self) -> dict[str, Any]:
+        return self._exec_kwargs
+
+
+# st.number_input
+class StreamlitNumberInputDifferentTypesError(LocalizableStreamlitException):
+    """Exception raised mixing floats and ints in st.number_input."""
+
+    def __init__(
+        self,
+        value: int | float | Literal["min"] | None,
+        min_value: int | float | None,
+        max_value: int | float | None,
+        step: int | float | None,
+    ):
+        value_type = None
+        min_value_type = None
+        max_value_type = None
+        step_type = None
+
+        error_message = "All numerical arguments must be of the same type."
+
+        if value:
+            value_type = type(value).__name__
+            error_message += "\n`value` has {value_type} type."
+
+        if min_value:
+            min_value_type = type(min_value).__name__
+            error_message += "\n`min_value` has {min_value_type} type."
+
+        if max_value:
+            max_value_type = type(max_value).__name__
+            error_message += "\n`max_value` has {max_value_type} type."
+
+        if step:
+            step_type = type(step).__name__
+            error_message += "\n`step` has {step_type} type."
+
+        super().__init__(
+            error_message,
+            value_type=value_type,
+            min_value_type=min_value_type,
+            max_value_type=max_value_type,
+            step_type=step_type,
+        )
+
+
+class StreamlitNumberInputInvalidMinValueError(LocalizableStreamlitException):
```

Comment:
I think I'd prefer something like `StreamlitWidgetValueBelowMinError`. "Invalid min value" sounds like the minimum is not allowed as opposed to a conflict between value and set boundaries. I'd also suggest a "Widget" error instead of a "NumberInput" error since I don't think the distinction between `st.number_input` and `st.slider` necessitates creating different classes here.

Reply:
My concern is future consequences and how easy it would be to change this pattern. If we aren't concerned about renaming exception classes and don't consider it a breaking change, then it's not as immediately critical. But if this creates inertia that will be hard to redirect as we create more exception classes and repeat this pattern on a larger scale, that's my primary concern that I'm hoping to clarify, especially as it relates to documentation.